[1단계 - 지하철 노선도 미션] 파노(전환오) 미션 제출합니다

jho2301 opened the pull request:
안녕하세요. 현아님.

지하철 미션으로 PR 드려요.

고민하고 있는 부분 질문드려요!

감사합니다.

## 요구사항

### 회원기능

- [x] 사용자는 회원 가입을 할 수 있다.

  - [x] 필요 정보: 이름, 이메일, 비밀번호, 비밀번호 확인 정보
  - [x] 비밀번호, 비밀번호 확인이 같은지 확인
  - [x] 실시간으로 회원가입 양식이 올바른지 사용자에게 알려준다.

- [x] 사용자는 가입한 계정으로 로그인을 할 수 있다.
  - [x] 로그인에 실패한 경우 실패한 이유를 확인할 수 있어야 한다.
- [x] 사용자는 로그인하여 지하철 노선도 관리 페이지에 접근할 수 있다.
- [x] 회원가입 성공시 자동 로그인 구현된다.

### 역 관리 기능

- [x] 사용자는 지하철 역을 추가할 수 있다.
  - [x] 역 이름은 2자 이상 20자 이하의 한글이어야 한다
  - [x] 역 이름은 숫자 포함 하지 않는다.
  - [x] 역 이름은 공백이 있을 수 없다.
- [x] 사용자는 지하철 역을 삭제할 수 있다.
- [x] 사용자는 등록되어 있는 전체 지하철 역 목록을 조회할 수 있다.

### 노선 관리 기능

- [x] 사용자는 지하철 노선을 추가할 수 있다.

  - 필요 정보: 노선 이름, 상행역, 하행역, (최초 상행역과 하행역 구간의)거리, (최초 상행역과 하행역 구간 통행에 걸리는)시간, 색상
  - [x] 노선 이름은 2자 이상 10자 이하의 한글이어야 한다.
  - [x] 노선 이름은 숫자 포함 하지 않는다.
  - [x] 노선 이름은 공백이 있을 수 없다.
  - [x] 상행선역을 설정하면 하행선 선택창에서는 해당 역이 존재하지 않는다.
  - [x] 상행역 설정 전에는 하행선역에 접근 불가히다.

  - [x] 상행역, 하행역: 기존에 등록되어 있는 지하철 역 목록 중에서 선택

- [x] 사용자는 지정되어 있는 10가지 색상 중 한 색상을 선택할 수 있다.
  - [x] 다른 노선에서 사용한 색은 선택할 수 없다.
- [x] 사용자는 등록되어 있는 지하철 노선을 삭제할 수 있다.
- [x] 사용자는 등록되어 있는 전체 지하철 노선 목록을 조회할 수 있다.

### 구간 관리 기능

- [x] 사용자는 특정 노선의 전체 구간 목록을 확인할 수 있다.
- [x] 사용자는 특정 지하철 노선에 구간을 추가할 수 있다.
  - [x] 하나의 역은 여러 개 노선에 중복되어 포함될 수 있음
  - [x] 역과 역 사이에 새로운 역 추가 가능
  - [x] 기존 구간의 거리보다 긴 거리의 사이 구간을 추가할 수 없다.
  - [x] 상행역에서는 해당 노선에 존재하는 역만 노출한다.
  - [x] 하행역에서는 전체 역 리스트 중 해당 노선의 역을 제외한 역들을 노출한다.
  - [x] 상행역 설정 전에는 하행역에 접근 불가하다.
  - 하나의 노선에서 갈래길은 생길 수 없음
- [x] 사용자는 노선에 등록되어 있는 구간을 삭제할 수 있다.

## 데모페이지

데모앱: https://react-subway-map.herokuapp.com/

## 웹 VSCode 환경

https://github1s.com/jho2301/react-subway-map/tree/step1


## 고민

1.  서비스로직을 모두 커스텀훅에 일임했습니다. 그리고 커스텀훅을 도메인 별로 나눴습니다. 여러 시도를 해보고 있는데 일임하는게 적절했는지 그리고  적절하게 분리했는지 피드백 받고 싶습니다.
    1. 커스텀훅으로 분리해도되나 가장 고민이 컸던 form컴포넌트의 훅을 커스텀훅으로 분리했는데, 컴포넌트가 몰라도되는 부분( setForm이나 validation 등의 구현부)이 있어서, 모든 로직을 커스텀훅으로 관리해도 되겠다 판단했습니다. 이렇게 관리하면 유지보수시 UI관련 코드만 볼지, 서비스로직 관련 코드만 볼 것인지 명확해질 것이라 예상했습니다.
    2. 에러 발생 시, 이를 어떻게 커스텀훅에서 컴포넌트로 전달해야할지는 좀 더 고민해야할 것 같습니다. (에러모달을 띄운다거나 할 때)
2.  커스텀훅에서 커스텀훅을 불러오는 부분이 있는데 커스텀훅끼리 알고있어도 괜찮을까요? 걱정되는건 순환참조  때문인데요. 1번을 잘 분리하면 해결되는 문제인지, 아니면 다른 방법을 강구해야할지 고민입니다,
3. [Best Practices for REST API Error Handling](https://www.baeldung.com/rest-api-error-handling-best-practices) 아티클을 읽어보니까 서버 에러케이스 분류하는 일을 에러코드로 받는게 좋았을 것 같은데 이번에 못해서 아쉽습니다. 현업에서는 어떻게 진행하나요?
4. 서비스로직 전부를 커스텀훅에서 작성해서 테스트를 훅테스트로 진행했습니다. 테스트를 기능 구현 후에 붙였는데 훅테스트할 때 TDD는 어떻게 하면 좋을까요? 지금 코드는 assert문이 구체적인 구현에 의존적인 것같습니다. (일부러 리액트쿼리 mutation이나 isSuccess를 리턴합니다)
5. 리액트쿼리 + 리덕스툴킷 사용했는데 적절하게 사용했는지 피드백 받고 싶습니다

woowapark committed:
Update README.md

ddongule committed:
chore: init Project with cra - typescript

ddongule committed:
chore: Storybook 세팅

ddongule committed:
feat: Global Style 적용

jho2301 committed:
chore: 스토리북 라우터 세팅

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>

jho2301 committed:
feat: NavigationBar 컴포넌트 마크업 작성

jho2301 committed:
feat: LoginPage 마크업 작성

jho2301 committed:
feat: SignUp 페이지 마크업 구현

jho2301 committed:
feat: 역 관리 페이지 마크업 작성

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>

jho2301 committed:
feat: 노선관리 페이지 마크업 작성

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>

jho2301 committed:
feat: 구간관리 페이지 마크업 작성

jho2301 committed:
test: 사용자가 회원가입을 할 수 있다

jho2301 committed:
feat: 사용자는 회원 가입을 할 수 있다.

jho2301 committed:
feat: 백엔드 API BASE_URL 상수화

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>

ddongule committed:
feat: 사용자는 가입한 계정으로 로그인 할 수 있다.

ddongule committed:
feat: PrivateRoute 구현

ddongule committed:
feat: 회원가입 성공시 자동 로그인 구현된다.

ddongule committed:
temp: Station 관리 Service 함수 추가함(API)

jho2301 committed:
chore: pages/singup 디렉토리 이름 변경

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>

jho2301 committed:
feat: API BASE_URL을 선택할 수 있다.

jho2301 committed:
feat: 선택된 API 이름 우하단에 표시

jho2301 committed:
fix: NavigationBar z-index 이슈

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>

jho2301 committed:
feat: 로그아웃을 할 수 있다

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>

jho2301 committed:
feat: API를 선택할 수 있는 페이지로 라우팅 기능

jho2301 committed:
feat: 사용자는 등록되어 있는 전체 지하철 역 목록을 조회할 수 있다

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>

jho2301 committed:
fix: 네비게이션바에 컨텐츠가 겹치는 현상

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>

jho2301 committed:
feat: 사용자는 지하철 역을 삭제할 수 있다.

ddongule committed:
feat: 사용자는 등록되어 있는 전체 지하철 노선 목록을 조회할 수 있다

ddongule committed:
feat: 사용자는 지하철 노선을 조회, 추가, 삭제할 수 있다.

jho2301 committed:
feat: 사용자는 노선에 등록되어 있는 구간을 삭제할 수 있다.

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>
Co-authored-by: JO YUN HO <bedro27@gmail.com>

jho2301 committed:
feat: 사용자는 특정 지하철 노선에 구간을 추가할 수 있다.

jho2301 committed:
fix: 노선 업데이트

jho2301 committed:
docs: 기능요구사항 추가

- 세부예외처리 사항

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>
Co-authored-by: JO YUN HO <bedro27@gmail.com>

jho2301 committed:
feat: 실시간으로 회원가입 양식이 올바른지 사용자에게 알려준다.

jho2301 committed:
feat: 로그인, 회원가입 폼 실시간 검증

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>

jho2301 committed:
feat: 에러 바운더리, 서스펜스 컴포넌트 적용

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>

jho2301 committed:
feat: 미완성 form submit button disabled 처리

jho2301 committed:
feat: 상행역 설정 전에는 하행선역에 접근 불가히다.

jho2301 committed:
feat: 상행선역을 설정하면 하행선 선택창에서는 해당 역이 존재하지 않는다

jho2301 committed:
feat: 상행역 설정 전에는 하행역에 접근 불가하다.

jho2301 committed:
feat: 구간 역 리스트 filtering

- 상행역에서는 해당 노선에 존재하는 역만 노출한다.
- 하행역에서는 전체 역 리스트 중 해당 노선의 역을 제외한 역들을 노출한다.

jho2301 committed:
feat: 기존 구간의 거리보다 긴 거리의 사이 구간을 추가할 수 없다.

jho2301 committed:
feat: 역 이름 유효성 검사

- 역 이름은 2자 이상 20자 이하의 한글이어야 한다
- 역 이름은 숫자 포함 하지 않는다
- 역 이름은 공백이 있을 수 없다.

jho2301 committed:
feat: 노선 이름 유효성 검사

- 노선 이름은 2자 이상 10자 이하의 한글이어야 한다.
- 노선 이름은 숫자 포함 하지 않는다.
- 노선 이름은 공백이 있을 수 없다.

ddongule committed:
feat: 노선에 등록된 역은 삭제할 수 없다.

jho2301 committed:
Merge branch 'step1' of https://github.com/ddongule/react-subway-map into step1

jho2301 committed:
feat: 상수화

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>

jho2301 committed:
chore: mock server 세팅

jho2301 committed:
test: 사용자는 로그인 할 수 있다.

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>

jho2301 committed:
test: 사용자는 로그아웃 할 수 있다

jho2301 committed:
test: 사용자는 로그아웃 할 수 있다.

jho2301 committed:
test: 사용자는 지하철 역을 추가할 수 있다.

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>

jho2301 committed:
test: 사용자는 지하철 역을 삭제할 수 있다.

jho2301 committed:
refactor: 테스트 가독성 개선

ddongule committed:
refactor: 사용자는 지정되어 있는 10가지 색상 중 한 색상을 선택할 수 있다.

ddongule committed:
test: useLine Hook 테스트

ddongule committed:
test: useLogin Hook 테스트

ddongule committed:
feat: 다른 노선에서 사용한 색은 선택할 수 없다.

jho2301 committed:
refactor: APISelect 페이지 커스텀훅 분리

jho2301 committed:
refactor: 사용하지 않는 모듈 import 제거

jho2301 committed:
refactor: 사용하지 않는 모듈 import 제거

jho2301 committed:
fix: input 길이가 의도된 너비를 벗어나는 현상

jho2301 committed:
fix: 글로벌 스타일 적용되지 않는 현상

jho2301 committed:
fix: login 에러메세지 제대로 나오지 않는 현상

jho2301 committed:
Merge branch 'step1' of https://github.com/jho2301/react-subway-map into step1

jho2301 committed:
refactor: 커스텀훅 form 로직 분리

Co-authored-by: Min Kyung Kang <mnk918p@gmail.com>

jho2301 committed:
fix: 테스트코드 정상화

jho2301 committed:
refactor: Form 커스텀훅 form 프로퍼티를 반환하지 않음

- form객체로 접근이 가능하다
- form객체로 접근하는 것이 더 명시적이다.

HyeonaKwon commented on src/components/@mixins/InputWithAlertText/AlertText/AlertText.tsx:
```diff
@@ -0,0 +1,22 @@
+  );
+};
+
+export default AlertText;
```
저는 컴포넌트명 보고 alert 안에 들어가는 text 컴포넌트인 줄 알았네요! ErrorText 나 ErrorMessageText 같이 에러에 대한 텍스트라는 것을 명시하는건 어떨까요??

jho2301 replied:
valid한 경우 invalid한 경우 두가지를 나타내고 있으므로 ValidationText로 바꿔보겠습니다.

HyeonaKwon commented on src/components/@mixins/InputWithAlertText/InputWithAlertText.tsx:
```diff
@@ -0,0 +1,32 @@
+  extends InputHTMLAttributes<HTMLInputElement>,
+    AlertTextProps {}
+
+const InputWithAlertText: VFC<InputWithAlertTextProps> = ({
```
props 에 따라 AlertText 를 보여줄건지 말건지를 받고 있는데 ~WithAlertText 라는건 항상 같이 보여준다는 의미로 보입니다!

jho2301 replied:
렌더 조건이 input value인데 키보드 입력하면 텍스트가 항상 나오므로 with라고 써줬던 것 같습니다.  이런 의도에도 부적절할지 궁금합니다

jho2301 replied:
위 코멘트의 컴포넌트와 같은 방식으로 네이밍하면 어떨까 생각해서 ValidationInput이 라고 변경해보았습니다.

HyeonaKwon commented on src/components/ErrorBoundary/ErrorBoundary.tsx:
```diff
@@ -0,0 +1,41 @@
+    const { error } = this.state;
+
+    if (error) {
+      return <h1>에러가 발생했습니다.</h1>;
```
이왕 사용하는거 좀 더 예쁜 UI 이면 좋겠네요 :)

jho2301 replied:
개선해보도록 하겠습니다

HyeonaKwon commented on src/components/LineManagement/LineAddModal/LineAddModal.tsx:
```diff
@@ -0,0 +1,123 @@
+            onChange={({ target }) => setUpStationId(Number(target.value))}
+          >
+            <Suspense fallback={true}>
+              {(stations.data as Station[]).map((station) => (
```
stations 의 타입이 Station[] 이어야하는거 아닌가요??

jho2301 replied:
감사합니다! 리액트쿼리의 쿼리객체이므로 stationsQuery가 더 알맞은 네이밍 같습니다

HyeonaKwon commented on src/components/LineManagement/LineAddModal/LineAddModal.tsx:
```diff
@@ -0,0 +1,123 @@
+                selectedColor={form.color}
+                backgroundColor={lineColor}
+                onClick={() => setColor(lineColor)}
+                disabled={lines.data?.some((line) => line.color === lineColor)}
```
요것도 마찬가지로 lines 이면 line[] 배열이어야할 것으로 보입니다.

jho2301 replied:
lineQuery로 수정하겠습니다!

HyeonaKwon commented on src/hooks/useLine.ts:
```diff
@@ -0,0 +1,54 @@
+import { QUERY } from '../constants/API';
+
+const useLine = () => {
+  const { accessToken } = useLogin();
```
커스텀 훅끼리 결합도를 가지고 있는 것은 좋지 않아보입니다! 순수하지도 않구요.
accessToken 이 필요한 함수에서 파라미터로 accessToken 을 받는게 좋을 것 같네요!

jho2301 replied:
사용하는 컴포넌트에서 조합하도록 해보겠습니다!

jho2301 replied:
모든 컴포넌트에서 accessToken를 불러오게 했는데, 때문에  응집도가 떨어진 것 같습니다. 바로 selector로 불러오게 해주거나 방법을  찾아봐야겠네요.

HyeonaKwon commented on src/hooks/useLineAddForm.ts:
```diff
@@ -0,0 +1,70 @@
+
+  const availableDownStations = stations.isSuccess
+    ? stations.data.filter((station) => station.id !== form.upStationId)
+    : [];
```
useStation 에서 isSuccess 인지에 따라 알맞은 stations 를 반환해줬다면 사용처에서 분기처리를 할 필요가 없었겠죠?

HyeonaKwon commented on src/hooks/useSection.ts:
```diff
@@ -0,0 +1,97 @@
+    addSection,
+    deleteSection,
+    updateCurrentSection,
+    isAddSectionSuccess,
+    isDeleteSectionSuccess,
```
요거는 왜 반환하는거죠..? 함수 내에서 반환해야하지 않을까요??

jho2301 replied:
리액트쿼리  mutation 객체가 addSection, deleteSection 함수 바깥에 있어서 그렇습니다. 
함수에서 mutation객체 자체를 내보내는 방법이 떠오르는데 이게 나을지 고민이네요.. 
좀 더 고민해보도록하겠습니다.

HyeonaKwon commented on src/service/auth.ts:
```diff
@@ -0,0 +1,28 @@
+  }
+
+  return response.json();
+};
```
service 레이어랑 hook 레이어랑 구분을 둔 기준은 뭘까요??

jho2301 replied:
제가 작성한 커스텀훅들은 서비스레이어의 일부라고 생각합니다. 다만 리액트에 와서 특별히 관리되는 컨벤션이 보여서 (함수명 prefix로 use를 붙이는) 파일도 다른 파일와 구분되도록  했습니다.  지금 생각해보면 서비스 디렉토리 내부에 배치하는게 더 직관적인 것 같습니다.

HyeonaKwon commented on src/tests/useLine.test.tsx:
```diff
@@ -0,0 +1,45 @@
+      expect(result.current.isDeleteLineSuccess).toBeTruthy()
+    );
+  });
+});
```
벨리데이션 쪽들도 추가되면 좋을 것 같네요!

HyeonaKwon replied:
사실 테스트는 정상 상황보다 예외 상황을 테스트하는게 더 효율적이니까요!

jho2301 replied:
엣지케이스를 좀 더 추가하도록 하겠습니다

HyeonaKwon reviewed (changes requested):
안녕하세요 파노님! 미션 구현하시느라 고생 많으셨어요 :)
피드백과 아래 내용 확인 부탁드려요!

- hooks 들에서 정제된 데이터를 제공해야 하는데 response 자체를 제공하고 있어 해당 데이터를 받아오는게 실패했는지 성공했는지를 사용처에서 계속 분기치는 현상이 발생하고 있습니다. 요런것들을 줄여보면 좋을 것 같아요! 
그리고 hook 들 서로가 결합도가 강해보입니다. 이렇게 되면 순수하지도 않을 뿐만 아니라 하나의 hook 이 잘못되면 여러 hook 들이 모두 잘못되는 현상이 발생할 수 있겠죠??
- hook 과 service 가 분리되면서 redux 의 보일러플레이트가 많은 현상이 더 증폭되어 보입니다. 하나의 api 가 추가되어도 건들여야하는 파일은 더 많아지는거죠. 요런 것들도 한 번 생각해보시면 좋을 것 같아요.
- 저희팀은 서버쪽에서 상황에 맞는 http status code 와 error code, error message 를 응답값으로 보내줍니다. 또는 message 는 프론트가 가지고 있고 서버에서 보내주는 error code 에 매핑되는 메세지를 사용합니다. 단순 메세지 변경은 프론트만 배포해도 충분하니까요!
- hook 을 UI 단위로 나누는게 아니라 도메인 단위로 나눴다면 더 재사용성이 높았을 것 같네요! form 으로 특정지었다보니 해당 form 에서밖에 사용할 수 없는 현상이 발생하죠?
- 테스트 코드에 예외 처리나 벨리데이션 관련 테스트를 추가한다면 지금보다는 더 풍부한 테스트 코드가 되었을 것 같네요!

jho2301 commented:
피드백 중 동의되지 않는 부분에 대해 다시 의견 듣고싶습니다.

- hook 을 UI 단위로 나누는게 아니라 도메인 단위로 나눴다면 더 재사용성이 높았을 것 같네요! form 으로 특정지었다보니 해당 form 에서밖에 사용할 수 없는 현상이 발생하죠?
:  처음에 도메인 단위로  커스텀훅을 만들었다가 bd1434b 에서 form관련 코드를 분리했습니다.   예상보다 깔끔하게 코드가 분리되는 경험을 했는데 이는 form코드와 다른 코드의 응집도가 낮았기 때문이라고 생각합니다.  이때는 코드를 분리하는게 맞지 않나 생각했는데 , 같은 도메인영역에 있어 같이 뭉쳐놓는게 좋았을까요?  커스텀훅 재사용성의 이점이 더 크기 떄문인지 궁금합니다

jho2301 commented:
피드백 중 고민했던 부분이 있어서 다시 의견 여쭙고 싶습니다.

hook 을 UI 단위로 나누는게 아니라 도메인 단위로 나눴다면 더 재사용성이 높았을 것 같네요! form 으로 특정지었다보니 해당 form 에서밖에 사용할 수 없는 현상이 발생하죠?
: 처음에 도메인 단위로 커스텀훅을 만들었다가 bd1434b 에서 form관련 코드를 분리했습니다. 이유는 form관련 로직만 수정하려고 하는 상황에 코드가 번잡해 많이 어려움을 겪어서 였고, 예상보다 깔끔하게 코드가 분리되는 경험을 했는데 이는 form코드와 다른 코드의 연관성이 낮았기 때문이라고 생각합니다. 이때는 form도 하나의 도메인으로 보고, 코드를 분리하는게 맞지 않나 생각했는데 , 같은 도메인 영역에 있어 같이 뭉쳐놓는게 좋았을까요? 커스텀훅 재사용성의 이점이 더 크기 떄문인지 궁금합니다

jho2301 committed:
feat: hooks 디렉토리 위치 변경

- service/hook

jho2301 committed:
refactor: 네이밍 변경

- AlertText컴포넌트 => ValidationText
- InputWithAlertText => ValidationInput
- stations => stationsQuery

jho2301 committed:
test: test renderHook함수 추상화

jho2301 committed:
refactor: reactQuery 로직 추상화

- 커스텀 훅간의 의존성 제거

jho2301 committed:
refactor: useSection 코드 개선

jho2301 committed:
feat: 에러페이지 컴포넌트 추가

jho2301 committed:
test: 역 이름은 2자 이상 20자 이하의 한글이어야 한다.

jho2301 committed:
test: 역 이름은 숫자를 포함하지 않는다

jho2301 committed:
test: 역 이름은 공백이 있을 수 없다.

jho2301 committed:
test: 노선 이름은 2자 이상 10자 이하의 한글이어야 한다.

jho2301 committed:
test: 노선 이름은 숫자를 포함하지 않는다.

jho2301 committed:
test: 노선 이름은 공백을 포함하지 않는다.

jho2301 committed:
test: 상행선역을 설정하면 하행선 선택창에서는 해당 역이 존재하지 않는다.

jho2301 commented:
추가로 리액트쿼리 API 사용하는 코드들 추상화 진행했습니다.  해당 과정으로 커스텀훅간 결합도를 낮출 수 있었습니다

HyeonaKwon commented on src/service/queries/line.ts:
```diff
@@ -0,0 +1,22 @@
+  return useMutation((lineId: LineId) =>
+    requestDeleteLine(lineId, accessToken)
+  );
+};
```
`useLineAddMutation`, `useLineDeleteMutation` 함수명이 너무 애매하네요🥲

jho2301 replied:
의도는 리액트쿼리 네이밍 그대로 이어받는거였는데 어렵네요..ㅜ

HyeonaKwon commented:
https://github.com/woowacourse/react-subway-map/pull/25#issuecomment-862098531
아 파노님! 맞습니다. 다만 도메인 별로 데이터를 저장하고 벨리데이션 했다면 재사용할 수 있는 상황이 있지 않았을까 싶었어요!
예를 들어, line 안에 station이 있다면 line form 에서는 station hook 을 사용할 수도 있으니까요!

HyeonaKwon commented:
queries 안에 있는 함수명들은 다시 고민해보는게 좋겠네요! 머지 진행해도 될까요??

jho2301 commented:
@HyeonaKwon 
기간 지난 이후로 귀찮게 해서 죄송합니다 ㅎㅎㅜㅜ
주신 피드백들은 2단계에서 계속 고민해보도록하겠습니다! 4개월간 감사했습니다

jho2301 commented:
# 학습로그

## 리액트쿼리

기존의 전역상태관리 라이브러리는 서버의 상태를 클라이언트와 동기화하는데 많은 노력을 들여야했다.
최신화전략, 캐싱전략, 에러, 로딩상태에 관한 홴들링 전략을 모두 작성해주어야했는데
리액트쿼리는 관련한 설정과 동작을 추상화하여 사용자가 편하게 서버상태를 동기화, 조작하는데 도움을 준다.

기본적으로 리액트쿼리 함수(useQuery, useMutation)에 key와 fetch콜백함수를 전달하는데,  fetch함수는 실제로 서버에 요청을 하는 함수, key는 각 서버요청에 대한 고유한 key를 리액트쿼리가 저장하는 것이다. 리액트쿼리는 key를 기준으로 캐시가 유효한지, 요청이 중복으로들어가는 경우를 파악해 요청을 하지 않는다거나 최적화를 진행한다.

이외에도 리액트쿼리는 refetch전략에 대해 최적화를 자체적으로 진행하였는데, 
대표적으로 
New instances of the query mount (새로운 쿼리객체가 마운트되었을 때)
The window is refocused(브라우저 창이 다시 포커스되었을 때)
The network is reconnected.(네트워크가 끊겼다가 다시 연결되었을 때)
The query is optionally configured with a refetch interval.(옵션으로 refetch간격이 설정되었을 때)
refetch를 진행한다.


## 백엔드/프론트엔드간 에러핸들링 컨벤션

에러가 발생했을 때  유저의 서비스 사용성 증진과 디버깅을 용이하게 하고자 할 때, 서버에서 에러상황을 세부적으로 클라이언트에 전달해주는 것이 가장 우선적으로 선행되어야할 것이다.

사용자의 잘못에서 기원한 에러일 경우,  유저는 자신이 어떤 에러를 발생시켰는지 피드백을 받을 수 있고,
디버깅의 목적일 경우, 어떤 에러가 발생했는지 디테일하게 로깅데이터를 봄으로써 신속하고 정확하게 이슈를 해결할 수 있다.

기본적으로는 http status code를 통해 에러의 종류를 클라이언트에게 알려줄 수 있다.

주요 코드 분류로

200번대: 성공
300번대: 리다이렉션
400번대: 클라이언트 에러
500번대: 서버 에러

를 나타낸다. 

주요 에러 응답코드는

400 Bad Request 
401 Unauthorized 
403 Forbidden 
404 Not Found 
500 Internal Server Error 
503 Service Unavailable 

가 있다.

하지만 세부적인 에러 상황은 알 수 없다. 
이런 상황을 해결하는 방법을 표준화하기 위해 RFC문서가 정해졌다.

[RFC7807](https://datatracker.ietf.org/doc/html/rfc7807)에서 

type – 에러를 카테고리화하는 URI 식별자
title – 사람이 읽을 수 있는 간단한 에러 내용
status – HTTP응답코드 (optional)
detail – 에러에 대한 상세설명
instance – 어떤 URI에서 에러가 발생했는지 

5개 부분으로 구성된 스키마를 규정했다.


가장 핵심적인 방법은 백엔드와 프론트엔드가 사전에 에러 코드(타입)를 합의한다는 것이다.

HyeonaKwon reviewed (approved):
귀찮게 하다니요! ㅎㅎ 파노님도 고생 많으셨습니다 🎉
